src/remote_tables/provider.rs

```diff
@@ -175,4 +214,17 @@ impl TableProvider for RemoteTable {
 
         Ok(plan)
     }
+
+    fn supports_filter_pushdown(
+        &self,
+        filter: &Expr,
+    ) -> Result<TableProviderFilterPushDown> {
+        // TODO: add caching of filter to expr mapping (here and in scan)?
+        if self.filter_expr_to_sql(filter).is_none() {
+            return Ok(TableProviderFilterPushDown::Unsupported);
+        }
+        // TODO: make this Inexact just to be on the safe side? The downside is that DF won't pushdown
+        // any LIMIT clause to the `scan` then, since it and WHERE are not commutable
+        Ok(TableProviderFilterPushDown::Exact)
```

Comment:
If we can only convert some expressions to the remote table's language, will DataFusion just try to push just those down? In other words, if some expressions are un-push-downable, does that ruin the entire pushdown?

Re: exact/inexact, I'd err on the side of exact and making `filter_expr_to_sql` return None if we're not sure that the expression will filter everything cleanly. We could also make `filter_expr_to_sql` itself return Exact/Inexact/Unsupported.

Re: caching, it's probably fine, we do this planning once per query, so I don't really anticipate it being the bottleneck.

Reply:
Yes, DataFusion optimizer is [smart enough](https://github.com/apache/arrow-datafusion/blob/master/datafusion/optimizer/src/filter_push_down.rs#L736-L756) to pushdown as much as possible and keep what it needs to ensure result correctness. It seems to split the filters up by the `AND`'s, since then it can still retain benefit from partial filter pushdown. 

Here's an example; for the following query
```sql
SELECT "date field", c FROM staging.remote_table WHERE (a > 2 OR c = 'two') AND (b < 5.0) LIMIT 2
```
the calls to `supports_filter_pushdown` and `scan` are as follows
```bash
supports_filter_pushdown is Exact: staging.remote_table.a > Int32(2) OR staging.remote_table.c = Utf8("two")
supports_filter_pushdown is Unsupported: CAST(staging.remote_table.b AS Float64) < Float64(5)
supports_filter_pushdown is Exact: staging.remote_table.a > Int32(2) OR staging.remote_table.c = Utf8("two")
supports_filter_pushdown is Exact: staging.remote_table.a > Int32(2) OR staging.remote_table.c = Utf8("two")
supports_filter_pushdown is Unsupported: CAST(staging.remote_table.b AS Float64) < Float64(5)
supports_filter_pushdown is Exact: staging.remote_table.a > Int32(2) OR staging.remote_table.c = Utf8("two")
scan received filter: a > Int32(2) OR c = Utf8("two")
final query: SELECT "b", "c", "date field" FROM "source table" WHERE a > 2 OR c = 'two'
```
The second filter is not shipable since there is no match for the CASE node in the filter expression visitor.

On the other hand, for
```sql
SELECT "date field", c FROM staging.remote_table WHERE (a > 2 OR c = 'two') AND (a < 5) LIMIT 2
```
I see
```bash
supports_filter_pushdown is Exact: staging.remote_table.a > Int32(2) OR staging.remote_table.c = Utf8("two")
supports_filter_pushdown is Exact: staging.remote_table.a < Int32(5)
supports_filter_pushdown is Exact: staging.remote_table.a > Int32(2) OR staging.remote_table.c = Utf8("two")
supports_filter_pushdown is Exact: staging.remote_table.a < Int32(5)
supports_filter_pushdown is Exact: staging.remote_table.a > Int32(2) OR staging.remote_table.c = Utf8("two")
supports_filter_pushdown is Exact: staging.remote_table.a < Int32(5)
scan received filter: a > Int32(2) OR c = Utf8("two")
scan received filter: a < Int32(5)
final query: SELECT "c", "date field" FROM "source table" WHERE a > 2 OR c = 'two' AND a < 5 LIMIT 2
```
You can see that not only have both filters been pushed down, the `LIMIT` clause was too.

As for caching, the number of visitor invocations per query per filter is 3-5, but the AST of the filter is bound to be very shallow so for now I think it's ok not to cache it. We can easily change that later on.